File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Ros/RosBridgeMessageOperation.swift
//
//  RosBridgeMessageOperation.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/11/1.
//

import Foundation

enum RosBridgeMessageOperation: String, Codable {
    case subscribe
    case publish
    case unsubscribe
    case callService = "call_service"
    case serviceResponse = "service_response"
}

/// RosBridgetClientのメッセージレスポンスの種別を判別するために使用する.
/// 判別した後、`publish`, `serviceResponse`毎にデコード処理を行う.
enum HandleRosBridgeMessageOperation {
    case publish
    case serviceResponse

    init?(_ op: RosBridgeMessageOperation) {
        switch op {
        case .publish:
            self = .publish
        case .serviceResponse:
            self = .serviceResponse
        default:
            return nil
        }
    }
}

struct HandleRosBridgeMessageResponse: Codable {
    var op: RosBridgeMessageOperation

    enum CodingKeys: String, CodingKey {
        case op
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/rosbridge/RosBridgeMessageOperation.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.rosbridge

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * [RosBridge Protocol](https://github.com/RobotWebTools/rosbridge_suite/blob/ros2/ROSBRIDGE_PROTOCOL.md)
 * の操作種別
 */
@Serializable
enum class RosBridgeMessageOperation {
    subscribe,
    publish,
    unsubscribe,
    @SerialName("call_service")
    callService,
    @SerialName("service_response")
    serviceResponse
}
